### PR TITLE
Parallelize a few operations in the FE_Q_Base constructor.

### PR DESCRIPTION
**Pull request by bangerth**

### CODE REVIEW

File: source/fe/fe_q_base.cc [hunk 1]
```diff
@@ -456,18 +456,18 @@ FE_Q_Base<PolynomialType,dim,spacedim>::initialize (const std::vector<Point<1> >
...
-  // finally fill in support points on cell and face
-  initialize_unit_support_points (points);
-  initialize_unit_face_support_points (points);
-
-  // reinit constraints
-  initialize_constraints (points);
+  // Finally fill in support points on cell and face and initialize
+  // constraints. All of this can happen in parallel
+  Threads::TaskGroup<> tasks;
+  tasks += Threads::new_task ([&]() { initialize_unit_support_points (points); });
```
Comment (masterleinad) [inline source/fe/fe_q_base.cc]: Seems you like the new possibility to have one line brace enclosed blocks. :smiley:

Comment (drwells) [reply source/fe/fe_q_base.cc]: I am surprised that astyle is happy with that one.

Comment (masterleinad) [reply source/fe/fe_q_base.cc]: Recently, we allowed having brace-enclosed one-liners. The main idea at that point was to allow for initializing variables, but we could only do this by allowing brace-enclosed one-liners in general.
Actually, I am not opposed to using it like that.

Comment (bangerth) [reply source/fe/fe_q_base.cc]: It's not a brace-enclosed initializer here, though. Or am I missing something in the discussion?

The alternative here would have been to write
```
  Threads::new_task (&FE_Q_Base<dim,spacedim>::initialize_unit_support_points,
                                  this,
                                  points);
```
but that just seems more complicated for no real gain.

Comment (masterleinad) [reply source/fe/fe_q_base.cc]: > It's not a brace-enclosed initializer here, though. Or am I missing something in the discussion?

No, it's not a brace-enclosed initializer. What I wanted to say is that `astyle` can't differentiate between a brace-enclosed initializer and anything else that is written within braces in one line.
We just don't get more granularity using `astyle`.
I am not sure if we should (try to) avoid using this style for anything but brace-enclosed initializers but in this case I am totally fine with it.

**Comment by bangerth** [comment]
Rebased. Let's see what happens now.

**Comment by masterleinad** [comment]
It's still the same:
```
base/task_11.debug: DIFF failed. ------ First 8 lines of numdiff/diff output:
----------------
##3       #:5   <== 5
##3       #:5   ==> 8
@ Absolute error = 3.0000000000e+0, Relative error = 6.0000000000e-1
----------------
##5       #:5   <== 1
##5       #:5   ==> 8
@ Absolute error = 7.0000000000e+0, Relative error = 7.0000000000e+0
base/task_11.debug: DIFF failed. ------ First 50 lines diff output:
*** /home/bob/source/tests/base/task_11.with_threads=on.output	Mon Aug  7 17:28:25 2017
--- output	Wed Apr  4 02:32:09 2018
***************
*** 1,8 ****
  
  DEAL::* default thread limit for tests: 5
! DEAL::max concurrent running: 5 should be: 5
  DEAL::* now with thread limit 1:
! DEAL::max concurrent running: 1 should be: 1
  DEAL::* now with thread limit 2:
! DEAL::max concurrent running: 2 should be: 2
  DEAL::OK
--- 1,8 ----
  
  DEAL::* default thread limit for tests: 5
! DEAL::max concurrent running: 8 should be: 5
  DEAL::* now with thread limit 1:
! DEAL::max concurrent running: 8 should be: 1
  DEAL::* now with thread limit 2:
! DEAL::max concurrent running: 8 should be: 2
  DEAL::OK
make[3]: *** [task_11.debug/diff] Error 1
make[2]: *** [CMakeFiles/task_11.debug.diff.dir/all] Error 2
make[1]: *** [CMakeFiles/task_11.debug.diff.dir/rule] Error 2
make: *** [task_11.debug.diff] Error 2
```

**Comment by tjhei** [comment]
> It's still the same:

This is weird. I assume this means that we run the code in this PR when executing this test even though we don't use any FEs. I guess TBB doesn't allow us to change the thread limits while threads are being used, so the test fails.

**Comment by kronbichler** [comment]
> I assume this means that we run the code in this PR when executing this test even though we don't use any FEs.

That would mean it runs in the global code section due to some static variable somewhere? @bangerth can you reproduce this on your machine?

**Comment by tjhei** [comment]
Maybe one of the mapping classes runs this code? The test might succeed if we start main() and sleep for a few seconds...

I am scared of possible side-effects like this for offloading stuff to separate threads.

**Comment by bangerth** [comment]
Hm, how bizarre. Let me see if I can reproduce this and figure out where this code is called from.

**Comment by bangerth** [comment]
This happens here:
```
template <int dim, int spacedim>
MappingQGeneric<dim,spacedim>
StaticMappingQ1<dim,spacedim>::mapping = MappingQGeneric<dim,spacedim>(1);
```
which calls
```
template <int dim, int spacedim>
MappingQGeneric<dim,spacedim>::MappingQGeneric (const unsigned int p)
  :
  polynomial_degree(p),
  line_support_points(this->polynomial_degree+1),
  fe_q(dim == 3 ? new FE_Q<dim>(this->polynomial_degree) : nullptr),
```
Anyone got an idea how to address this? I think the general idea to let the `FE_Q` constructor do a few things in parallel is useful. I guess `static` global variables are not, apparently. Suggestions?

**Comment by kronbichler** [comment]
> Anyone got an idea how to address this?

I don't like the `fe_q` member of `MappingQGeneric`. It is only used for `adjust_quad_dof_index_for_face_orientation` which in turn is filled in `source/fe/fe_q_base.cc` over here: https://github.com/dealii/dealii/blob/master/source/fe/fe_q_base.cc#L987

I guess it would make sense to just duplicate that information as it probably just as easy/difficult to maintain as to take the indirection over `FE_Q`. A solution in the long run would be to get rid of as many global variables as possible.

**Comment by tjhei** [comment]
> A solution in the long run would be to get rid of as many global variables as possible.

agreed.

@bangerth is the failing test caused by these tasks still running in the background or is this caused by the fact the we can not limit the number of threads if we already had some running in the past (and exited)? In other words, can you fix the test by waiting at the start of main? 

Maybe our logic in ``set_thread_limit`` is incorrect:
see https://github.com/dealii/dealii/blob/5ad27a86c866dd9354c76c7736c5a5d71e1a210a/source/base/multithread_info.cc#L148
```
  // Initialize the scheduler and destroy the old one before doing so
  static tbb::task_scheduler_init dummy (tbb::task_scheduler_init::deferred);
  if (dummy.is_active())
    dummy.terminate();
  dummy.initialize(n_max_threads);
```
Could it be that executing a task before this code runs for the first time means a different default``task_scheduler_init`` is created, so we can not terminate the old one using the static variable? A fix would then be to call ``set_thread_limit`` inside your new block of code before you run the tasks...

So, to me it looks like you are not only breaking this test but every piece of code that tries to change the number of threads...

**Comment by bangerth** [comment]
On 04/04/2018 09:40 AM, Timo Heister wrote:
> @bangerth <https://github.com/bangerth> is the failing test caused by 
> these tasks still running in the background or is this caused by the 
> fact the we can not limit the number of threads if we already had some 
> running in the past (and exited)? In other words, can you fix the test 
> by waiting at the start of main?

The patch waits for the tasks to finish. But the TBB has apparently 
already created the threads to runs these tasks and then refuses to get 
rid of them again when we later call `set_thread_limit()`.

> Maybe our logic in |set_thread_limit| is incorrect:
> see 
> https://github.com/dealii/dealii/blob/5ad27a86c866dd9354c76c7736c5a5d71e1a210a/source/base/multithread_info.cc#L148
> 
> |// Initialize the scheduler and destroy the old one before doing so 
> static tbb::task_scheduler_init dummy 
> (tbb::task_scheduler_init::deferred); if (dummy.is_active()) 
> dummy.terminate(); dummy.initialize(n_max_threads); |
> 
> Could it be that executing a task before this code runs for the first 
> time means a different default|task_scheduler_init| is created, so we 
> can not terminate the old one using the static variable?

I think this is correct.


> A fix would 
> then be to call |set_thread_limit| inside your new block of code before 
> you run the tasks...

We'd have to do that in every piece of code where we run tasks because 
we can't know whether it's going to be called from a static constructor 
before main() starts. What's more, `set_thread_limit()` takes an 
argument, but how would one choose that in some random place of the library?


> So, to me it looks like you are not only breaking this test but every 
> piece of code that tries to change the number of threads...

That is correct.

**Comment by masterleinad** [comment]
> Could it be that executing a task before this code runs for the first time means a different defaulttask_scheduler_init is created, so we can not terminate the old one using the static variable? A fix would then be to call set_thread_limit inside your new block of code before you run the tasks...

For me, that even doesn't fix the problem.

**Comment by drwells** [comment]
I agree with @kronbichler: the simplest solution is to get rid of the `FE_Q` member of `MappingQGeneric` and store the table directly instead.

I believe this would be easy to fix. Currently `FE_Q_Base::initialize_quad_dof_index_permutation` populates the table. We could move this function into `internal::FE_Q_BaseImplementation` or wherever and then just call it in `FE_Q_Base` as a nonmember function and then again in `MappingQGeneric` (which would now own a table instead of a pointer to an `FE_Q`). We would still have to duplicate `adjust_quad_dof_index_for_face_orientation` but that is three assertions and one table lookup.

**Comment by bangerth** [comment]
I had looked into this, and my previous two patches were actually preparatory for this, but it's going to take me a bit of time to get to it. Let's put this into WIP status for now.

I think the bigger issue is that that may not actually solve the problem. Several of the functions in `mapping_q_generic.cc` actually create a local variable of type `FE_Q`. In other words, whenever we create the `static` Q1 mapping variable, if the constructor calls any of the functions that create the `FE_Q`, then we're back to the same point as before. The `static` `FE_Q` member variable we're talking about here may simply be the first `FE_Q` variable we create, and that's why we're talking about it, but it may not be the last.

**Comment by tjhei** [comment]
> whether it's going to be called from a static constructor
> before main() starts. What's more, `set_thread_limit()` takes an
> argument, but how would one choose that in some random place of the library?

We just need to make sure that https://github.com/dealii/dealii/blob/master/source/base/multithread_info.cc#L213
is executed before we create any tasks. Using ``invalid_unsigned_int`` is the correct thing to do.

We could place a call to something like ``ensure_set_thread_limit_has_been_called`` into places where we create tasks:
```
void ensure_set_thread_limit_has_been_called() 
{
  static bool done = false;
  if (!done) done=true;
  else MultithreadInfo::set_thread_limit (numbers::invalid_unsigned_int);
}
struct DoOnce
  {
    DoOnce ()
    {
        ensure_set_thread_limit_has_been_called();
    }
  } do_once;
```

**Comment by bangerth** [comment]
Oh, I didn't remember this code.

Are you saying that the place in `multithread_info.cc` executes the constructor of the `DoOnce` class, but unfortunately does so *after* we create the static mapping variable?

There are ways that affect the order in which global variables are initialized. If we could ensure that the `do_once` variable in the snippet you linked to is initialized *before* the mapping variable is created, do you think that would work?

**Comment by tjhei** [comment]
> unfortunately does so after we create the static mapping variable?

correct.

> variable in the snippet you linked to is initialized before the mapping variable is created, do you think that would work?

Yes. Which is simple enough to do by essentially calling it whenever you try to create a task (see the blob of code I posted). Do you want me to try this?

**Comment by bangerth** [comment]
I can do it, but probably not in the next few days if you want to try.

But if you wrote a PR against this PR, I'd merge it while I'm traveling :-)

**Comment by tjhei** [comment]
> But if you wrote a PR against this PR, I'd merge it while I'm traveling :-)

This works for me: https://github.com/bangerth/dealii/pull/3

I am not sure if there is a better way to achieve this...

**Comment by masterleinad** [comment]
I would expect that the overhead is only checking for a `bool` to be true or false. That is certainly really cheap in comparison to basically everything we would be doing else.

**Comment by tjhei** [comment]
> overhead is only checking for a bool

the function call is probably more costly than the check, but I agree that this is unlikely to be a problem.

**Pull request merged**